core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -0,0 +1,639 @@
+   */
+  protected String getParentPath(String path) {
+    // Root does not have a parent.
+    if (isRoot(path)) {
+      return null;
+    }
+    int separatorIndex = path.lastIndexOf(PATH_SEPARATOR);
+    if (separatorIndex < 0) {
+      return null;
+    }
+    return path.substring(0, separatorIndex);
+  }
+
+  /**
+   * Checks if the key is the root.
+   *
+   * @param key ufs path including scheme and bucket
+   * @return true if the key is the root, false otherwise
+   */
+  protected boolean isRoot(String key) {
+    return PathUtils.normalizePath(key, PATH_SEPARATOR).equals(
+        PathUtils.normalizePath(getRootKey(), PATH_SEPARATOR));
+  }
+
+  /**
+   * Gets the child name based on the parent name.
+   *
+   * @param child the key of the child
+   * @param parent the key of the parent
+   * @return the child key with the parent prefix removed, null if the parent prefix is invalid
+   */
+  protected String getChildName(String child, String parent) {
+    if (child.startsWith(parent)) {
+      return child.substring(parent.length());
+    }
+    LOG.error("Attempted to get childname with an invalid parent argument. Parent: {} Child: {}",
+        parent, child);
+    return null;
+  }
+
+  /**
+   * Get suffix used to encode a directory.
+   *
+   * @return folder suffix
+   */
+  protected abstract String getFolderSuffix();
+
+  /**
+   * Get object listing for given path.
+   *
+   * @param path pseudo-directory path excluding header and bucket
+   * @param recursive whether to request immediate children only, or all descendants
+   * @return chunked object listing
+   * @throws IOException if a non-alluxio error occurs
+   */
+  protected abstract ObjectListingResult getObjectListing(String path, boolean recursive)
+      throws IOException;
+
+  /**
+   * Get full path of root in object store.
+   *
+   * @return full path including scheme and bucket
+   */
+  protected abstract String getRootKey();
+
+  /**
+   * Lists the files in the given path, the paths will be their logical names and not contain the
+   * folder suffix. Note that, the list results are unsorted.
+   *
+   * @param path the key to list
+   * @param recursive if true will list children directories as well
+   * @return an array of the file and folder names in this directory
+   * @throws IOException if an I/O error occurs
+   */
+  protected UnderFileStatus[] listInternal(String path, boolean recursive) throws IOException {
+    // if the path not exists, or it is a file, then should return null
+    if (!isDirectory(path)) {
+      return null;
+    }
+    path = stripPrefixIfPresent(path);
+    path = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    path = path.equals(PATH_SEPARATOR) ? "" : path;
+    Map<String, Boolean> children = new HashMap<>();
+    ObjectListingResult chunk = getObjectListing(path, recursive);
+    chunk = chunk.getNextChunk();
+    if (chunk == null) {
+      return null;
```

Comment:
chunk is null if 
- if there list didn't find anything with prefix path (initially)
- or, a chunk was processed and there are no more chunks remaining

the 2 places where getObjectListing is called were passing in a path which had been normalized with PATH_SEPARATOR. I changed it so that the implementation normalizes it if not already so. the return should be null if path is a file. i need to add test cases for this class.